Group loading of translations for integrations to reduce executor jobs at startup

bdraco opened the pull request:
- [x] split bugfix into https://github.com/home-assistant/core/pull/110748
- [x] split of part of this into https://github.com/home-assistant/core/pull/110732
- [x] It would be nice to merge https://github.com/home-assistant/core/pull/110130 first since it will conflict here.

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Group loading of translations for integrations

This is a cleanup to followup on https://github.com/home-assistant/core/pull/96906#issuecomment-1946834787 to avoid a thundering heard of executor jobs at startup.

Although it wasn't the goal, this also happens fix #73148 because since we load in bootstrap now it became very obvious where the problem was since bootstrap tests didn't fully mock the integration.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #73148
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

bdraco committed:
Group loading of translations for integrations

bdraco committed:
cleanups

bdraco committed:
it really loads them all

bdraco committed:
less complex

bdraco committed:
less complex

bdraco committed:
reduce tasks

bdraco committed:
revert unneeded changes

bdraco committed:
collapse

bdraco committed:
cleanups

bdraco committed:
tweak

bdraco committed:
group load

bdraco committed:
remove debug

bdraco committed:
fixes

bdraco committed:
fix mocking

bdraco commented on homeassistant/helpers/translation.py:
```diff
@@ -135,17 +135,17 @@ def _merge_resources(
     components: set[str],
     category: str,
-) -> dict[str, dict[str, Any] | str]:
+) -> dict[str, str | dict[str, Any]]:
```
```suggestion
) -> dict[str, dict[str, Any] | str]:
```

bdraco commented on homeassistant/helpers/translation.py:
```diff
@@ -135,17 +135,17 @@ def _merge_resources(
 
 def build_resources(
-    translation_strings: dict[str, dict[str, Any]],
+    translation_strings: dict[str, dict[str, str | dict[str, Any]]],
```
```suggestion
    translation_strings: dict[str, dict[str, dict[str, Any] | str]],
```

bdraco committed:
Apply suggestions from code review

bdraco commented on homeassistant/helpers/translation.py:
```diff
@@ -465,30 +452,62 @@ def async_setup(hass: HomeAssistant) -> None:
             component,
         )
-        await _async_load_state_translations_to_cache(hass, language, component)
+        await cache.async_load(language, {component})
```
This call will almost never happen now since it will already be loaded unless the integration is loaded late

bdraco commented on homeassistant/helpers/translation.py:
```diff
@@ -258,17 +264,20 @@ async def _async_load(self, language: str, components: set[str]) -> None:
             language,
-            ", ".join(components),
+            components,
         )
         # Fetch the English resources, as a fallback for missing keys
         languages = [LOCALE_EN] if language == LOCALE_EN else [LOCALE_EN, language]
 
         integrations: dict[str, Integration] = {}
-        domains = list({loaded.partition(".")[0] for loaded in components})
+        domains = {loaded.partition(".")[0] for loaded in components}
```
The conversion to list was not needed since `async_get_interations` takes an iterable

bdraco commented on homeassistant/bootstrap.py:
```diff
@@ -673,6 +674,10 @@ async def _async_set_up_integrations(
         requirements.async_load_installed_versions(hass, needed_requirements),
         "check installed requirements",
     )
+    hass.async_create_background_task(
```
Maybe a comment here to explain why would be good.  We want to load the translations as soon as possible in a single job to avoid a thundering herd that all has to wait on the translations lock

bdraco committed:
comment

bdraco committed:
comment

emontnemery commented on homeassistant/bootstrap.py:
```diff
@@ -673,6 +674,17 @@ async def _async_set_up_integrations(
         requirements.async_load_installed_versions(hass, needed_requirements),
         "check installed requirements",
     )
+    # Start loading translations for all integrations we are going to set up
+    # in the background so they are ready when we need them. This avoids a
+    # lot of waiting for the lock and a thundering herd of tasks trying to
+    # load the same translations at the same time as each integration is loaded.
+    # We do not wait for this since as soon as the task runs it will
+    # hold the lock and if anything is fast enough to wait for the lock,
+    # loading will be done by the time it gets to it.
```
Please clarify this comment, it's not clear what lock is being referred to.

bdraco replied:
adjusted comment

emontnemery reviewed (commented):
The translation code is a bit difficult to understand, would it be possible to cut this PR down into smaller pieces, for example do the event filtering in a separate PR.

bdraco committed:
clarify which lock

bdraco committed:
Add filter to translation events to avoid creating tasks

We no longer create tasks when the the task will be a no-op

This is not so useful on its own but it supports
#110674 and was split to make the review smaller

bdraco committed:
Add filter to translation events to avoid creating tasks

We no longer create tasks when the the task will be a no-op

This is not so useful on its own but it supports
#110674 and was split to make the review smaller

bdraco commented:
> The translation code is a bit difficult to understand, would it be possible to cut this PR down into smaller pieces, for example do the event filtering in a separate PR.

Split into https://github.com/home-assistant/core/pull/110732

bdraco committed:
Merge branch 'dev' into translation

bdraco committed:
Merge branch 'dev' into translation_no_task

bdraco committed:
only fire if lang changes

bdraco committed:
lint

bdraco committed:
Merge branch 'translation_no_task' into translation

bdraco committed:
race safe

bdraco committed:
Merge branch 'translation_no_task' into translation

bdraco committed:
Merge branch 'dev' into translation

emontnemery commented on homeassistant/helpers/translation.py:
```diff
@@ -135,17 +135,17 @@ def _merge_resources(
-        if category in translation_strings[component]
-        and translation_strings[component][category] is not None
+        if (component_strings := translation_strings.get(component))
+        and (category_strings := component_strings.get(category))
```
This changes the semantics somewhat since we now check for truthiness instead of for not None. I'm not sure if that matters?

bdraco replied:
We would get KeyError if the integration had failed to load and we skipped it

emontnemery replied:
OK

emontnemery commented on homeassistant/helpers/translation.py:
```diff
@@ -263,17 +264,20 @@ async def _async_load(self, language: str, components: set[str]) -> None:
         for domain, int_or_exc in ints_or_excs.items():
             if isinstance(int_or_exc, Exception):
-                raise int_or_exc
+                _LOGGER.warning(
+                    "Failed to load integration for translation: %s", int_or_exc
+                )
+                continue
```
It somehow feels like it would be nice to keep this change of behavior outside this optimization PR. Feel free to move it to a separate PR or not. Regardless of which, there should be a test which tests for the new behavior and fails with the old behavior.

bdraco replied:
Added coverage for partial failure 85327a1251509923065d0ac3ed97482c2f71dd1d

emontnemery commented on homeassistant/helpers/translation.py:
```diff
@@ -388,31 +392,15 @@ async def async_get_translations(
     """
     if integrations is None and config_flow:
         components = (await async_get_config_flows(hass)) - hass.config.components
+    elif integrations is not None:
+        components = set(integrations)
```
Why is this case removed?

bdraco replied:
Its moved to line 398. I think its just the formatting??

bdraco committed:
coverage for partial failure

bdraco committed:
fix patch

bdraco commented:
drafting this to pull it out of the review queue as https://github.com/home-assistant/core/pull/110748 should merge first

bdraco committed:
Merge branch 'dev' into translation

PiotrMachowski commented:
@bdraco I have checked if these changes introduced any regression regarding #96906 and I think everything should be fine 👍

bdraco commented:
Thanks @PiotrMachowski
